th/Categorifier/TH.hs

```diff
@@ -37,14 +40,30 @@ import Data.Map (Map)
 import qualified Data.Map as Map
 import Data.Maybe (catMaybes, fromMaybe)
 import Data.These (These (..))
-import Language.Haskell.TH hiding (TyVarBndr, conP, reifyType)
+import Language.Haskell.TH hiding (TyVarBndr (..), conP, reifyType)
 import qualified Language.Haskell.TH as TH
 import PyF (fmt)
 
 #if MIN_VERSION_template_haskell(2, 17, 0)
 type TyVarBndr flag = TH.TyVarBndr flag
+
+pattern PlainTV :: TH.Name -> flag -> TyVarBndr flag
+pattern PlainTV n flag = TH.PlainTV n flag
+
+pattern KindedTV :: TH.Name -> flag -> TH.Kind -> TyVarBndr flag
+pattern KindedTV n flag k = TH.KindedTV n flag k
+
+{-# COMPLETE PlainTV, KindedTV #-}
 #else
 type TyVarBndr _flag = TH.TyVarBndr
+
+pattern PlainTV :: TH.Name -> flag -> TyVarBndr flag
+pattern PlainTV n _flag = TH.PlainTV n
+
+pattern KindedTV :: TH.Name -> flag -> TH.Kind -> TyVarBndr flag
+pattern KindedTV n _flag k = TH.KindedTV n k
```

Comment:
We don't need these patterns to be bidirectional, do we? They can be just
```haskell
pattern KindedTV n _flag k -> TH.KindedTV n k
```
But we can leave the 2.17 ones bidirectional.

Reply:
Oh, nevermind ... I got the unidirectional ones backward. I guess they would be like
```haskell
pattern KindedTV :: TH.Name -> TH.Kind -> TyVarBndr flag
#if MIN_VERSION_template_haskell(2, 17, 0)
pattern KindedTV n k <- TH.KindedTV n _flag k
#else
pattern KindedTV n k = TH.KindedTV n k
#endif
```
where the _old_ can be bidirectional, but the new can't.

Reply:
I was hoping `pattern KindedTV :: TH.Name -> flag -> TH.Kind -> TyVarBndr flag` for both versions, but I'm not sure if that's possible. I think your approach is probably OK, so I'll do it